New Plugin: ws281x

mzanetti opened the pull request:
Reopening this as it was accidentally merged. For reference: #1

hermanndetz committed:
Initial commit

hermanndetz committed:
Added files missing from previous commit

hermanndetz committed:
Added details about rpi_ws281x lib to README.md

hermanndetz committed:
Removed plugininfo.h

This file is generated by nymea-plugininfocompiler.

hermanndetz committed:
Extended README.md

Added section on hardware connection (pins + activation using
raspi-config)

hermanndetz committed:
Added note about scons to README.md

hermanndetz committed:
Fixed crossbuilder build

Removed scons dependency
Fixed include path for lib in submodule
Added translations file

simonseres commented:
I built the plugin on my pi zero again. There was some issue. May @hermanndetz has a quick idea on why that happens.
```
nymea@nymea-pi0:~/nymea-plugin-ws281x $ git submodule init
nymea@nymea-pi0:~/nymea-plugin-ws281x $ git submodule update
Submodule path 'rpi_ws281x': checked out '31f668cc637cd96e132b611db82cbe9de9838366'
nymea@nymea-pi0:~/nymea-plugin-ws281x $ mkdir build
nymea@nymea-pi0:~/nymea-plugin-ws281x $ cd build
nymea@nymea-pi0:~/nymea-plugin-ws281x/build $ qmake ..
WARNING: Failure to find: plugininfo.h
nymea@nymea-pi0:~/nymea-plugin-ws281x/build $ make
nymea-plugininfocompiler ../integrationpluginws281xspi.json --output plugininfo.h
g++ -c -pipe -std=c++11 -g -O2 -Wall -W -D_REENTRANT -fPIC -DQT_NO_DEBUG -DQT_PLUGIN -DQT_GUI_LIB -DQT_CORE_LIB -I../../nymea-plugin-ws281x -I. -I../rpi_ws281x -isystem /usr/include/nymea -isystem /usr/include/arm-linux-gnueabihf/qt5 -isystem /usr/include/arm-linux-gnueabihf/qt5/QtGui -isystem /usr/include/arm-linux-gnueabihf/qt5/QtCore -I. -I/usr/lib/arm-linux-gnueabihf/qt5/mkspecs/linux-g++ -o integrationpluginws281xspi.o ../integrationpluginws281xspi.cpp
nymea-plugininfocompiler ../integrationpluginws281xspi.json --extern extern-plugininfo.h
g++ -c -pipe -std=c++11 -g -O2 -Wall -W -D_REENTRANT -fPIC -DQT_NO_DEBUG -DQT_PLUGIN -DQT_GUI_LIB -DQT_CORE_LIB -I../../nymea-plugin-ws281x -I. -I../rpi_ws281x -isystem /usr/include/nymea -isystem /usr/include/arm-linux-gnueabihf/qt5 -isystem /usr/include/arm-linux-gnueabihf/qt5/QtGui -isystem /usr/include/arm-linux-gnueabihf/qt5/QtCore -I. -I/usr/lib/arm-linux-gnueabihf/qt5/mkspecs/linux-g++ -o ws281xspicontroller.o ../ws281xspicontroller.cpp
gcc -c -pipe -O2 -Wall -W -D_REENTRANT -fPIC -DQT_NO_DEBUG -DQT_PLUGIN -DQT_GUI_LIB -DQT_CORE_LIB -I../../nymea-plugin-ws281x -I. -I../rpi_ws281x -isystem /usr/include/nymea -isystem /usr/include/arm-linux-gnueabihf/qt5 -isystem /usr/include/arm-linux-gnueabihf/qt5/QtGui -isystem /usr/include/arm-linux-gnueabihf/qt5/QtCore -I. -I/usr/lib/arm-linux-gnueabihf/qt5/mkspecs/linux-g++ -o mailbox.o ../rpi_ws281x/mailbox.c
gcc -c -pipe -O2 -Wall -W -D_REENTRANT -fPIC -DQT_NO_DEBUG -DQT_PLUGIN -DQT_GUI_LIB -DQT_CORE_LIB -I../../nymea-plugin-ws281x -I. -I../rpi_ws281x -isystem /usr/include/nymea -isystem /usr/include/arm-linux-gnueabihf/qt5 -isystem /usr/include/arm-linux-gnueabihf/qt5/QtGui -isystem /usr/include/arm-linux-gnueabihf/qt5/QtCore -I. -I/usr/lib/arm-linux-gnueabihf/qt5/mkspecs/linux-g++ -o pwm.o ../rpi_ws281x/pwm.c
gcc -c -pipe -O2 -Wall -W -D_REENTRANT -fPIC -DQT_NO_DEBUG -DQT_PLUGIN -DQT_GUI_LIB -DQT_CORE_LIB -I../../nymea-plugin-ws281x -I. -I../rpi_ws281x -isystem /usr/include/nymea -isystem /usr/include/arm-linux-gnueabihf/qt5 -isystem /usr/include/arm-linux-gnueabihf/qt5/QtGui -isystem /usr/include/arm-linux-gnueabihf/qt5/QtCore -I. -I/usr/lib/arm-linux-gnueabihf/qt5/mkspecs/linux-g++ -o rpihw.o ../rpi_ws281x/rpihw.c
g++ -pipe -std=c++11 -g -O2 -Wall -W -dM -E -o moc_predefs.h /usr/lib/arm-linux-gnueabihf/qt5/mkspecs/features/data/dummy.cpp
/usr/lib/qt5/bin/moc -DQT_NO_DEBUG -DQT_PLUGIN -DQT_GUI_LIB -DQT_CORE_LIB --include /home/nymea/nymea-plugin-ws281x/build/moc_predefs.h -I/usr/lib/arm-linux-gnueabihf/qt5/mkspecs/linux-g++ -I/home/nymea/nymea-plugin-ws281x -I/home/nymea/nymea-plugin-ws281x/rpi_ws281x -I/usr/include/nymea -I/usr/include/arm-linux-gnueabihf/qt5 -I/usr/include/arm-linux-gnueabihf/qt5/QtGui -I/usr/include/arm-linux-gnueabihf/qt5/QtCore -I. -I/usr/include/c++/8 -I/usr/include/arm-linux-gnueabihf/c++/8 -I/usr/include/c++/8/backward -I/usr/lib/gcc/arm-linux-gnueabihf/8/include -I/usr/local/include -I/usr/lib/gcc/arm-linux-gnueabihf/8/include-fixed -I/usr/include/arm-linux-gnueabihf -I/usr/include ../integrationpluginws281xspi.h -o moc_integrationpluginws281xspi.cpp
g++ -c -pipe -std=c++11 -g -O2 -Wall -W -D_REENTRANT -fPIC -DQT_NO_DEBUG -DQT_PLUGIN -DQT_GUI_LIB -DQT_CORE_LIB -I../../nymea-plugin-ws281x -I. -I../rpi_ws281x -isystem /usr/include/nymea -isystem /usr/include/arm-linux-gnueabihf/qt5 -isystem /usr/include/arm-linux-gnueabihf/qt5/QtGui -isystem /usr/include/arm-linux-gnueabihf/qt5/QtCore -I. -I/usr/lib/arm-linux-gnueabihf/qt5/mkspecs/linux-g++ -o moc_integrationpluginws281xspi.o moc_integrationpluginws281xspi.cpp
/usr/lib/qt5/bin/moc -DQT_NO_DEBUG -DQT_PLUGIN -DQT_GUI_LIB -DQT_CORE_LIB --include /home/nymea/nymea-plugin-ws281x/build/moc_predefs.h -I/usr/lib/arm-linux-gnueabihf/qt5/mkspecs/linux-g++ -I/home/nymea/nymea-plugin-ws281x -I/home/nymea/nymea-plugin-ws281x/rpi_ws281x -I/usr/include/nymea -I/usr/include/arm-linux-gnueabihf/qt5 -I/usr/include/arm-linux-gnueabihf/qt5/QtGui -I/usr/include/arm-linux-gnueabihf/qt5/QtCore -I. -I/usr/include/c++/8 -I/usr/include/arm-linux-gnueabihf/c++/8 -I/usr/include/c++/8/backward -I/usr/lib/gcc/arm-linux-gnueabihf/8/include -I/usr/local/include -I/usr/lib/gcc/arm-linux-gnueabihf/8/include-fixed -I/usr/include/arm-linux-gnueabihf -I/usr/include ../ws281xspicontroller.h -o moc_ws281xspicontroller.cpp
g++ -c -pipe -std=c++11 -g -O2 -Wall -W -D_REENTRANT -fPIC -DQT_NO_DEBUG -DQT_PLUGIN -DQT_GUI_LIB -DQT_CORE_LIB -I../../nymea-plugin-ws281x -I. -I../rpi_ws281x -isystem /usr/include/nymea -isystem /usr/include/arm-linux-gnueabihf/qt5 -isystem /usr/include/arm-linux-gnueabihf/qt5/QtGui -isystem /usr/include/arm-linux-gnueabihf/qt5/QtCore -I. -I/usr/lib/arm-linux-gnueabihf/qt5/mkspecs/linux-g++ -o moc_ws281xspicontroller.o moc_ws281xspicontroller.cpp
rm -f libnymea_integrationpluginws281xspi.so
g++ -std=c++11 -z defs -Wl,-O1 -shared -o libnymea_integrationpluginws281xspi.so integrationpluginws281xspi.o ws281xspicontroller.o mailbox.o pwm.o rpihw.o moc_integrationpluginws281xspi.o moc_ws281xspicontroller.o  -lnymea /usr/lib/arm-linux-gnueabihf/libQt5Gui.so /usr/lib/arm-linux-gnueabihf/libQt5Core.so /usr/lib/arm-linux-gnueabihf/libGLESv2.so -lpthread /usr/lib/gcc/arm-linux-gnueabihf/8/libatomic.so    
/usr/bin/ld: ws281xspicontroller.o: in function `Ws281xSpiController::freeLEDString()':
/home/nymea/nymea-plugin-ws281x/build/../ws281xspicontroller.cpp:152: undefined reference to `ws2811_fini'
/usr/bin/ld: ws281xspicontroller.o: in function `Ws281xSpiController::initLEDString()':
/home/nymea/nymea-plugin-ws281x/build/../ws281xspicontroller.cpp:174: undefined reference to `ws2811_init'
/usr/bin/ld: ws281xspicontroller.o: in function `Ws281xSpiController::update()':
/home/nymea/nymea-plugin-ws281x/build/../ws281xspicontroller.cpp:209: undefined reference to `ws2811_render'
collect2: error: ld returned 1 exit status
make: *** [Makefile:163: libnymea_integrationpluginws281xspi.so] Error 1
nymea@nymea-pi0:~/nymea-plugin-ws281x/build $ 
```

hermanndetz commented:
I'll look into it

mzanetti commented:
Jenkins, test this please

nymea-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.nymea.io/job/nymea-plugin-ws281x-pull-request-tester/2/

nymea-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.nymea.io/job/nymea-plugin-ws281x/4/

hermanndetz commented:
I got a clue, what happens, but no fix yet:

Comparing these two targets from the Makefile:
`
ws281xspicontroller.o: ../nymea-plugin-ws281x/ws281xspicontroller.cpp ../nymea-plugin-ws281x/ws281xspicontroller.h \
		extern-plugininfo.h
	$(CXX) -c $(CXXFLAGS) $(INCPATH) -o ws281xspicontroller.o ../nymea-plugin-ws281x/ws281xspicontroller.cpp

mailbox.o: rpi_ws281x/mailbox.c 
	$(CC) -c $(CFLAGS) $(INCPATH) -o mailbox.o rpi_ws281x/mailbox.c
`

The ws281xspicontroller.o target has the correct path to the src directory (../nymea-plugin-ws281x/) in my case. But the target mailbox.o (and all others from the library) don't.

They are listed one after the other in the ws281xspi.pro file:

`
SOURCES += \
    integrationpluginws281xspi.cpp \
    ws281xspicontroller.cpp \
    rpi_ws281x/mailbox.c \
    rpi_ws281x/pwm.c \
    rpi_ws281x/rpihw.c
`

mzanetti commented:
I've set up jenkins testing for this repository. Hermann, please contact me for login credentials on jenkins if you want to see the build logs. That's the reference for getting it build in the end.

What I'm a bit unsure still, on what distributions and architectures should this be available? I've tried building it in crossbuilder for buster/armhf and it fails, as well as on raspbian/armhf natively on a raspberry pi.

(Also, please add `qt5-default` as build dependency. Our crossbuilder images have that preinstalled, but on a plain raspbian/debian it fails without that).

hermanndetz committed:
Build + docs fixes

Modifications to project file to fix build process (tested on x64, RPi
3+ and crossbuilder)
Added qt5-default to debian/control
Changed "git checkout" for "git clone" in README.md

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/nymea/nymea-plugin-ws281x?pullRequest=3) <br/>All committers have signed the CLA.

nymea-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.nymea.io/job/nymea-plugin-ws281x/7/

nymea-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.nymea.io/job/nymea-plugin-ws281x-pull-request-tester/3/

mzanetti commented:
Jenkins, test this please

nymea-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.nymea.io/job/nymea-plugin-ws281x-pull-request-tester/4/

mzanetti commented:
Jenkins, test this please

nymea-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.nymea.io/job/nymea-plugin-ws281x-pull-request-tester/5/
